### DISCUSSION TITLE
ImageMagick Gif Application Extension block seems to get in the way of NETSCAPE2.0 block

### DISCUSSION THREAD

**Discussion by discretegames** [ImageMagick/ImageMagick#4813] [Help] [answered]
Hello! 👋 I've been toying with ImageMagick and I'm super glad it exists as a tool. 😊 But I ran into an issue I haven't been able to figure out.

I made this gif:

![out](https://user-images.githubusercontent.com/35741644/152656400-30ae50e4-be51-4ea3-a2cb-f89ec1f0086e.gif)

with the command `magick -delay 50 *.png -loop 0 -layers Optimize out.gif` from these three images named `1.png 2.png 3.png`:
![1](https://user-images.githubusercontent.com/35741644/152656319-dcea4004-0a7e-4fe9-ba6c-8e65621c7a3d.png)
![2](https://user-images.githubusercontent.com/35741644/152656324-898413d6-7b8d-4536-a0aa-6a26599ce02e.png)
![3](https://user-images.githubusercontent.com/35741644/152656325-13bee0fd-42e9-415d-8f06-7eaf33aa2917.png)

It works fine and loops infinitely as expected in most places, but notably in Window's own Photo viewer it only loops once and stops at the final light gray frame: 

![image](https://user-images.githubusercontent.com/35741644/152656463-ba1e3cf8-7735-41d7-8be3-b212e74184ac.png)

I'm not certain but I think this this is because ImageMagick adds an an Application Extension block before the NETSCAPE2.0 one that encodes the number of loops (in fact it add ones before every frame, but I think only the first matters). You can see this on [this excellent gif explorer site](http://www.matthewflickinger.com/lab/whatsinagif/gif_explorer.asp):

![image](https://user-images.githubusercontent.com/35741644/152656693-c16d9d2a-8497-4c9b-8589-f6b5c157063c.png)

In fact when I manually removed the ImageMagick Application Extension block from the file, it started looping as expected in Windows Photos:

![out-edited](https://user-images.githubusercontent.com/35741644/152657274-3484e739-1f04-45a0-9032-6cee2effb338.gif)

[This site](http://www.vurdalakov.net/misc/gif/netscape-looping-application-extension) (one of [the few](http://www.matthewflickinger.com/lab/whatsinagif/animation_and_transparency.asp) I've been able to find about adding loops to gifs) says that

> Netscape Looping Application Extension is the most popular Application Extension Block that tells browser or other GIF viewer to loop the entire animated GIF file.
>
> This block must appear immediately after the global color table of the logical screen descriptor.

which are the instructions I'm guessing Windows Photos is following, hence the gif doesn't loop. Whereas other viewers presumably loop infinitely by default or detect any NETSCAPE2.0 blocks.

So my question is, is there any way to prevent ImageMagick from adding its Application Extension blocks to a gif, or to reorder them so the NETSCAPE2.0 one is first?

Then I think the gifs produced by ImageMagick will work in the default Windows 10 image viewer. (Though maybe really I should be complaining to Windows. This is not the first problem I've had with their image viewer 😆)

Thank you very much for taking the time to respond!

**Comment by urban-warrior** [comment, marked as answer]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/download/beta/ by sometime tomorrow.
